File: BinaryTreeRightSideView_dfs.java
//https://leetcode.com/submissions/detail/617859764/
//Time: O(n)
//Space: O(n)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    List<Integer> rightView;
    public List<Integer> rightSideView(TreeNode root) {
        int level = 0;
        rightView = new ArrayList<>();
        dfs(root, 0);
        return rightView;
    }
    
    private void dfs(TreeNode root, int level) {
        
        if(root == null) {
            return;
        }
        
        if(rightView.size() == level) {
            rightView.add(root.val);   
        }        
        
        dfs(root.right, level + 1);
        dfs(root.left, level + 1);
    }
    
}

File: binary-tree-right-side-view-dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# TC: O(n)
# SC: O(h)
class Solution:
    def __init__(self):
        self.result = []

    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []

        self.dfs(root, 0)

        return self.result

    def dfs(self, root, level):
        # base
        if root is None:
            return

        # logic
        if level == len(self.result):
            self.result.append(root.val)

        self.dfs(root.right, level + 1)
        self.dfs(root.left, level + 1)
